Refactor extract and profile transformation

hiker opened the pull request:
As discussed in #474 this small pull request starts to restructure some transformations - it basically moves all *Extract* and Profile transformations into the future directory structure (plus RegionTrans and TransformationError to psyir/transformations - the latter because I tried to avoid some circular references). Some comments and discussions:
1) I have not updated the TransInfo code. Imho this class does not offer a significant advantage over using the transformation name directly (especially now that we have agreed on a naming scheme), but with potentially dozens of transforms in separate files it would not require to read all those files to find transforms, which I think is pretty expensive (since the user will typically only want one or two transforms). So far *Extract and Profile transformation tests do not use TransInfo, so the tests (and examples) all work.

2) I have not touched psyir/nodes - afaik @arporter has started doing this in a separate branch, so I want to avoid conflicts. My ongoing work on extract has already moved them the nodes into the right subdirectory (and I will handle the conflict when Andy's branch is merged).

3) I've added a section to the developers guide about file name convention and directory structure. It's by far not complete, but should cover the changes I started to implement.

hiker commented:
Ready for first review.

One comment re circular reference: In profile_trans.py have to use:
```
from psyclone.psyir.transformations.transformation_error \
    import TransformationError
```
If I don't specify the transformation_error file, I get an import error when trying to import ProfileTrans:

- profile_trans.py imports TransformationError, so if I don't specify the filename, it must read __init__.py (in psyir.transformation).
- __init__.py imports (after a while) profile_trans  --> circle

So by using from ``psyclone.psyir.transformations.transformation_error import TransformationError`` there is no need to read __init__, which avoids the circular reference.

Why don't we see this with e.g. symbols? Because the order in which the classes are imported in symbols.__init__.py. If I change the order in transformation.__init__.py, I can also avoid this problem - but then the imports are not alphabetically sorted anymore.

So, it's either alphabetically sorting the imports in __init__.py, or using full filename in the subdirectories.

arporter commented on src/psyclone/psyir/transformations/region_trans.py:
```diff
@@ -0,0 +1,170 @@
+    # pylint: disable=abstract-method,arguments-differ
+    '''
+    This abstract class is a base class for all transformations that act
+    on a list of nodes. It gives access to a _validate function that
```
s/_validate/validate/

hiker replied:
Done.

arporter commented on src/psyclone/psyir/transformations/region_trans.py:
```diff
@@ -0,0 +1,170 @@
+        # Sanity check that we've not been passed the condition part of
+        # an If statement or the bounds of a Loop. If the parent node is
+        # a Loop of IfBlock then we can only accept a single Schedule.
+        # TODO #542 Once everything has a Schedule we can tidy this up
```
#542 has been done and this remains! My fault. Would you mind correcting along the lines of this comment?

hiker replied:
I hope I did the right thing there.

arporter replied:
Almost! You just need to tighten-up the second part of the check to make sure that the list has only one member and that that member is a Schedule. (Since e.g. the if-body and else-body nodes are both Schedules but it would be an error to supply both of them to a transformation.)

arporter replied:
Yes, thank you :-)

arporter commented on src/psyclone/psyir/transformations/transformation_error.py:
```diff
@@ -0,0 +1,51 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors I. Kavcic, Met Office,
```
Should probably just have Rupert and yourself as authors.

hiker replied:
OK - that is a problem with most restructures, I can't really tell who I should list. I put Rupert there (but not myself ... I only copy&pasted the file).

arporter replied:
Thanks.

arporter commented:
I need to think about the circular-dependency problem a bit more but that will have to wait until tomorrow...

hiker commented on doc/developer_guide/coding-style.rst:
```diff
@@ -127,3 +127,41 @@ Some important details:
+The directory structure of the PSyclone classes is as follows:
+
+domain:
+    This directory contains the various API specific classes.
```
Done.

hiker commented on doc/developer_guide/coding-style.rst:
```diff
@@ -127,3 +127,41 @@ Some important details:
+        The following domains are currently supported: `lfric`, `gocean`, `nemo`.
+
+        domain/API_NAME/transformations:
+            These directories contain for each API the api-specific transformations,
```
Done.

hiker commented on doc/developer_guide/coding-style.rst:
```diff
@@ -127,3 +127,41 @@ Some important details:
+        domain/API_NAME/transformations:
+            These directories contain for each API the api-specific transformations,
+            typically using one of the classes in psyir/transformations as base
+            class. Any transformation class should have the domain as postfix, and
```
Oops

hiker commented on doc/developer_guide/coding-style.rst:
```diff
@@ -127,3 +127,41 @@ Some important details:
+            These directories contain for each API the api-specific transformations,
+            typically using one of the classes in psyir/transformations as base
+            class. Any transformation class should have the domain as postfix, and
+            `Trans` as postfix (e.g. `GOceanExtractTrans`), and corresponding file
```
Done.

hiker commented on doc/developer_guide/coding-style.rst:
```diff
@@ -127,3 +127,41 @@ Some important details:
+    psyir/transformations
+        This directory contains all basic transformations, i.e. all transformations
+        that are either directly usable in any API, or are base classes for
+        api-specific transformations. Any transformation class should have `Trans`
```
Done.

hiker commented on doc/developer_guide/coding-style.rst:
```diff
@@ -127,3 +127,41 @@ Some important details:
+        This directory contains all basic transformations, i.e. all transformations
+        that are either directly usable in any API, or are base classes for
+        api-specific transformations. Any transformation class should have `Trans`
+        as postfix (e.g. `ExtractTrans`), and corresponding file name should end
```
Done.

hiker commented on examples/gocean/eg2/acc_prof_transform.py:
```diff
@@ -53,8 +53,8 @@ def trans(psy):
     '''
-    from psyclone.transformations import ProfileRegionTrans
-    proftrans = ProfileRegionTrans()
+    from psyclone.psyir.transformations import ProfileRegion
```
Yes ... and I also updated the example script to actually use this transform :)

hiker commented on src/psyclone/domain/gocean/transformations/gocean_extract_trans.py:
```diff
@@ -0,0 +1,98 @@
+    >>> from psyclone.domain.gocean.transformations import GOceanExtractTrans
+    >>> etrans = GOceanExtractTrans()
+    >>>
+    >>> # Apply GOceanExtractTransTrans transformation to selected Nodes
```
Good things come in pairs :)

hiker commented on src/psyclone/domain/lfric/transformations/lfric_extract_trans.py:
```diff
@@ -0,0 +1,99 @@
+
+
+class LFRicExtractTrans(ExtractTrans):
+    ''' Dynamo0.3 API application of ExtractTrans transformation \
```
I admit I am not sure about lfric vs dynamo.

hiker commented on src/psyclone/profiler.py:
```diff
@@ -44,7 +44,7 @@
 
 
-class Profiler(object):
+class Profiler():
```
Yes, it does. Only pylint with python2 complains, but I believe we agreed to make everything python3-clear.

arporter replied:
Sure.

hiker commented on src/psyclone/profiler.py:
```diff
@@ -291,13 +293,15 @@ def update(self):
         super(ProfileNode, self).update()
 
         # Get the parse tree of the routine containing this region
+        # pylint: disable=protected-access
```
No, I tried that (and tried it now again) - no such thing.

arporter replied:
Oh, my mistake. One day soon(ish) we won't need _ast anyway.

hiker commented on src/psyclone/psyir/transformations/__init__.py:
```diff
@@ -0,0 +1,53 @@
+
+# The entities in the __all__ list are made available to import directly from
+# this package e.g.:
+# from psyclone.psyir.transformations import ExtractRegion
```
Indeed.

hiker commented on src/psyclone/psyir/transformations/extract_trans.py:
```diff
@@ -0,0 +1,220 @@
+# -----------------------------------------------------------------------------
+# Authors I. Kavcic, Met Office
+
+'''This module contains the base class for extracting kernel parameter
```
Hmm - done, but technically that is not right at this stage - it only adds code to extract the kernel parameter (well, in the future) - creating a stand-alone application has not been started afaik.

arporter replied:
Oh yes, that's true but it is the ultimate aim so I suggest we live with it...

hiker commented on src/psyclone/psyir/transformations/extract_trans.py:
```diff
@@ -0,0 +1,220 @@
+            # this may be between an orphaned Directive (e.g. OMPDoDirective,
+            # ACCLoopDirective) and its ancestor Directive (e.g. ACC or OMP
+            # Parallel Directive) or within an OMPParallelDoDirective.
+            if node.ancestor(OMPParallelDirective) or \
```
Yep, and resolved.

arporter replied:
Great.

hiker commented:
I spent some time thinking if ExtractTrans should be abstract (i.e. no name or __str__), in which case we could just remove the functions and not add tests. But since I wasn't sure about the future, I just added the tests to cover the remaining lines in ExtractTrans.

Assuming travis is fine, ready for next review.

arporter commented on examples/check_examples:
```diff
@@ -157,6 +157,7 @@ cd gocean/eg2
 # We don't do the full compile because that requires a Fortran compiler
 # and the dl_esm_inf library
 ${PSYCLONE} -api "gocean1.0" -s ./acc_transform.py alg.f90
+${PSYCLONE} -api "gocean1.0" -s ./acc_prof_transform.py alg.f90
```
Thanks.

hiker commented on src/psyclone/psyir/transformations/region_trans.py:
```diff
@@ -0,0 +1,167 @@
+        # an If statement or the bounds of a Loop. If the parent node is
+        # a Loop or IfBlock then we can only accept a single Schedule.
+        if not isinstance(node_parent, Schedule) and \
+            not(isinstance(node_list[0], Schedule)):
```
Fixed.

hiker commented on src/psyclone/tests/extract_test.py:
```diff
@@ -35,20 +35,21 @@
 # Modified by A. R. Porter, STFC Daresbury Lab
```
Done. I was actually thinking the same ... but then the issue is that extract_test mostly uses gocean and dynamo to do the tests, which would then raise the question if these should be moved into domain/API/transformations ... which then makes it a pain to test all extract functionality with a single test.

arporter replied:
Strictly speaking we should be able to separate-out the PSyIR-only parts of the tests but I'm happy to leave it for now.

hiker commented on src/psyclone/tests/gocean0p1_transformations_test.py:
```diff
@@ -14,9 +14,8 @@
 from psyclone.configuration import Config
```
Done.

arporter replied:
Great.

hiker commented:
OK, I am all confused now. I have an email from you stating:
> Almost! You just need to tighten-up the second part of the check to make sure that the list has only one member and that that member is a Schedule. (Since e.g. the if-body and else-body nodes are both Schedules but it would be an error to supply both of them to a transformation.)

But I can't find that comment on github. Not sure what I am doing wrong.

Anyway, I spend the better part of two hours thinking about this test (since I actually originally had a test if the list is length 1, which I removed later), and I _believe_ the current comment in the code indicates that the test is indeed correct the way it is:
```
            # We've already checked for lists with len > 1 that contain a
            # Schedule above so if the first item is a Schedule then that's
            # all the list contains.
```
So I don't think the test if the list is len 1 is necessary: if the list has more than one element, then
- if there is a Schedule in there, the previous if-test will have triggered.
- if there is no schedule in the list, then it is sufficient to test the first element, which I am doing

If you still want me to add this additional test, I can do that - then one clarification: which type can node_list be? It is documented to be either a list or a Schedule, but some code seems to indicate that it is always a list:
```
        node_parent = node_list[0].parent
        prev_position = -1
        for child in node_list:
```
I tried giving it a schedule only, and it aborted in:
```
            prev_position = child.position
```
Should I change the docu that it must be a list??

arporter commented:
> OK, I am all confused now. I have an email from you stating:
> 
That's annoying - I did write that comment. Then I subsequently looked at the code in more detail and saw the comment you quoted and realised that everything was OK. I then deleted my comment but didn't realise that github had already emailed you :-( Long story short - you don't need to do anything :-)

hiker commented:
> moving some of the test files to match the restructuring of the source files

Sure ... not much work *sigh*  What was I thinking to agree to quickly do it 😡 ???? *sigh* But I think I got it all fixed now. And I won't volunteer to move any files around in the near future :)

arporter commented:
All tests pass/xfail, including with compilation.
All requested changes have been made.
This is ready to go once the conlicts have been sorted out. (Apologies for the delay and the resulting conflicts but I wanted to get the profiling-transformation fixes in ASAP for the Met Office.)

hiker commented:
Done - ready to go!

arporter commented:
I was all set to merge this but have just discovered that there are a lot of errors when building the User Guide. Please could you correct the names and import paths of the transformations in that document (and in the dev. guide if necessary).

hiker commented:
I completely missed the docs, thanks a lot!

I've updated the doc, and put two notes in that the location of the transformation is changing. We also kind of forgot the issue about TransInfo that I've mentioned above.I've documented that the new transforms can not be imported using TransInfo in the docs, added #620 and a TODO.

I also noticed that there were several scripts still importing TransformationError from psyclone.transformation, and not psyclone.psyir.transformation. That worked since psyclone.transformation imports TransformationError, but I believe I've updated all instances of this incorrect import.

Ready for another look.

arporter commented:
All documentation now builds OK (thanks).
All tests pass/xfail including with compilation.
Examples all run OK and look OK.
